### PR TITLE
Get ubuntu 16.04 on ppc to work!

### PR DESCRIPTION
**Pull request by doublej472**
This is actually ubuntu 16.04.02 (now 16.04.03), but it should be very similar for our purposes.

This whole PR is basically a conglomeration of [chef bento](https://github.com/chef/bento/) and [our old packer](https://github.com/osuosl/bento) repos, with fixes thrown everywhere to pass the openstack taster tests. So far it passes the security tests fine, but it has trouble mounting some filesystems from the volumes, which is why this is still a work in progress.

I'm making this PR early if anyone wants to look at what I've done to get ubuntu to cooperate (This took a lot of random changes to get working), and if anyone wants to use the images they are listed as `Ubuntu 16.04 LE doublejtest` and `Ubuntu 16.04 BE doublejtest` on openpower-controller.osuosl.org

This is still in desperate need of squashing, I'll get to it once volumes work completely.

### CODE REVIEW

**Comment by doublej472** [comment]
Some more info for reviewers:

The `scripts/common/minimize.sh` script was updated to work when no swapfile (or partition) existed, it won't fail anymore. It will also no longer fail when writing zero to the disks.

`scripts/common/sshd.sh` was completely overhauled so that it will work on nearly any system we give it to. Each directive is tested to exist, if it does then ensure that it has the default values, if not, write the value to the config.

`scripts/ubuntu/networking.sh` is interesting because we are actually appending kernel parameters to the bootloader so that we get an `eth0` interface. It will add the parameter to any bootloader it can find, regardless if it is actually used. This is because Ubuntu BE and LE use different bootloaders (Yaboot and GRUB2 respectively), we need to check for both configs. This script also manually disables the cloud-init networking (I could not get it working properly, booting would always fail an bringing up network interfaces), and configure the networks manually.

**Comment by knightsamar** [comment]
Here's the current openstack_taster output: 

LE: https://pastebin.osuosl.org/62466/
BE: https://pastebin.osuosl.org/62471/

The BE failure seems to be related to BTRFS being used on the two volumes (also remember @doublej472 mentioning it) but I always thought we had it formatted all volumes with ext4.

The PR looks good to me and I also think this is a very good sample to get running through the pipeline (almost ready) since it affects scripts which in turn affect other images.

**Comment by knightsamar** [comment]
Build fails for this with packer-pipeline. Looks like 16.04.3 came out and they deleted the 16.04.2 image, by the time we were deploying packer-pipeline.

``
==> qemu: Downloading or copying ISO
    qemu: Downloading or copying: http://cdimage.ubuntu.com/releases/16.04/release/ubuntu-16.04.2-server-powerpc.iso
    qemu: Error downloading: checksums didn't match expected: ea6744626bc9ec4833e76b7855c6e31f
==> qemu: ISO download failed.
Build 'qemu' errored: ISO download failed.
``

File: scripts/ubuntu/openstack.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+apt-get -y install pump cloud-utils cloud-init cloud-initramfs-growroot \
+    bash-completion
+
+# Ignore wardning about datasources
+mkdir -p /etc/cloud/cloud.cfg.d
+cat <<EOF >/etc/cloud/cloud.cfg.d/99-warnings.cfg;
+#cloud-config
+warnings:
+  dsid_missing_source: off
+EOF
```
Comment (ramereth) [inline scripts/ubuntu/openstack.sh]: Why is this needed?

File: scripts/ubuntu/vmware.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh -eux
+
+case "$PACKER_BUILDER_TYPE" in
+vmware-iso|vmware-vmx)
+    apt-get install -y open-vm-tools;
+    mkdir /mnt/hgfs;
+    echo "platform specific vmware.sh executed";
+esac
```
Comment (ramereth) [inline scripts/ubuntu/vmware.sh]: We don't need this.

**Comment by knightsamar** [comment]
Successfully testing this using packer_pipeline requires #26 and also #32 to be merged and then this branch rebased upon master.

Right now the images are successfully built and deployed but we cannot run openstack_taster because of a bug which is addressed by #32

File: scripts/ubuntu/openstack.sh (hunk 1 above)
Comment (doublej472) [inline scripts/ubuntu/openstack.sh]: Without this file, anytime you log into an instance you get a scary warning about the instance only being able to find Amazon EC2 datasources, and not openstack datasources that we have it configured to use.

File: scripts/ubuntu/networking.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# Fix ethernet address for network interfaces
+sed -i 's/auto\ en.*/auto\ eth0/' /etc/network/interfaces
+sed -i 's/iface\ en.*/iface\ eth0\ inet\ dhcp/' /etc/network/interfaces
+
+# Adding a 5 sec delay to the interface up, to make the dhclient happy
+echo "pre-up sleep 5" >> /etc/network/interfaces;
+
+# Disable cloud-init networking configuration
+mkdir -p /etc/cloud/cloud.cfg.d
+echo "network: {config: disabled}" > /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg
```
Comment (doublej472) [inline scripts/ubuntu/networking.sh]: Keeping cloud-init configuring networks was causing problems for me, the instances never got their networking set up correctly. Right now this PR sets an `eth0` interface to dhcp manually, but I can see now that this will cause problems when we get multiple interfaces. I'll work on getting this working without having to set this.

**Comment by doublej472** [comment]
@knightsamar when I looked at the [packer-pipeline jenkins](https://jenkins.osuosl.org/job/packer_pipeline/) I saw that job 34 was frozen on `start_processing_on_right_nodes` for 49 minutes. I've aborted it now, but in case you wanted to know.

**Comment by doublej472** [comment]
@ramereth this PR should now address everything you brought up, although there is one part I kept in. You can see why in a reply to one of your review comments. Networking seems to be working fine now, no need to explicitly disable cloud-init. This passes almost all of the `openstack_taster` tests (Ubuntu BE has trouble mounting some filesystems), and I believe it is ready.

File: scripts/ubuntu/openstack.sh (hunk 1 above)
Comment (ramereth) [inline scripts/ubuntu/openstack.sh]: Does this happen when you deploy the VM on an OpenStack cluster?

Comment (doublej472) [inline scripts/ubuntu/openstack.sh]: I believe so, I created the instances on the cluster when testing them.

Comment (ramereth) [inline scripts/ubuntu/openstack.sh]: Just to document it here. I'd recommend instead of turning off this warning to update the config to use the correct datasource. You can do something similar to what we did in the [old bento repo for Debian](https://github.com/osuosl/bento/blob/master/packer/scripts/debian/openstack.sh#L7-L17).

**Comment by doublej472** [comment]
@ramereth The ubuntu images now don't spit out the warning anymore, I had to take a different approach than what you recommended, but it seems to work better anyways. Essentially, I'm forcing the images to use OpenStack as a datasource and flat out disabling EC2.

This should be ready for review again, the uploaded `doublejtest` images are from the current version of this branch.

**Comment by knightsamar** [comment]
!taste

**Comment by knightsamar** [comment]
!taste

**Comment by ramereth** [comment]
[ubuntu 16.04 BE & LE template fix up to pass openstack_taster](https://trello.com/c/DGb18Fdt/487-ubuntu-1604-be-le-template-fix-up-to-pass-openstacktaster)

**Comment by knightsamar** [comment]
The actual build results are at: https://jenkins.osuosl.org/job/packer_pipeline/226/console

The failures for mitaka builds are in RT#29691. I am not able to figure how to fix it.
@ramereth / @doublej472 could you please look into that while I close the loose ends of this pipeline ?

**Comment by doublej472** [comment]
Part of it looks to be failing on Big Endian `BTRFS` filesystem mounting, which is a problem I brought up a while ago. This probably isn't something I can fix without some serious kernel magic.

As of writing this, I can't connect to [openpower-openstack.testing.osuosl.org](openpower-openstack.testing.osuosl.org), which is what I think you are testing on if I am to believe the logs. I can't even ping it from my workstation. It looks like this happened half way through the tests?

**Comment by ramereth** [comment]
@doublej472 that cluster is offline until we rebuild it. IMO we should only include production openstack clusters in this pipeline otherwise we're going to run into this problem. Or at least it needs to work around it in case they are offline gracefully.

**Comment by knightsamar** [comment]
I have started a run with all the latest fixes to the pipeline and the Mitaka images at https://jenkins.osuosl.org/job/packer_pipeline/238/console.

It should set a status on this PR and we should be able to know clearly what things to fix in the templates. This PR has now become so drenched in major changes to the pipeline that unless it's merged we cannot really expect to reliably test #33 or #36.

I won't be in until Monday evening (so logistically Tuesday) but I don't believe anything is pending fix on the Pipeline side for this PR to be tasted or merged. One major grouse is **!deploy** is still not automated but I can run that script manually once this PR is merged.

**Comment by knightsamar** [comment]
The build still fails on mitaka templates: https://pastebin.osuosl.org/92311/

**Comment by knightsamar** [comment]
The latest build got stuck at `KVM: Failed to create TCE table for liobn 0x80000001` for the `ubuntu-16.04-ppc64le` template. This is described and tracked at https://bugzilla.redhat.com/show_bug.cgi?id=1440619

Other status:

`centos-7.2-ppc64-openstack.json` build fails due to missing ISO.
centos-7.3-ppc64le-openstack.json:[linter:0, builder:0],
ubuntu-16.04-ppc64le-openstack.json:[linter:0] failed due ot the error above, 
centos-7.3-ppc64-openstack.json:[linter:0] was not yet built 
, ubuntu-16.04-ppc64-openstack.json:[linter:0]] was not yet built.

**Comment by doublej472** [comment]
@knightsamar Looks like it doesn't like the CDROM driver?

https://pastebin.osuosl.org/100316/

**Comment by knightsamar** [comment]
@doublej472 ok i will try removing it and rebuilding.

**Comment by knightsamar** [comment]
I am going to try looking at building using qemu-kvm directly and then changing what packer passes to it.

**Comment by doublej472** [comment]
@knightsamar I think I have the Ubuntu LE template working now, my commit should fix the cdrom error we were facing above. Manually running packer on openpower2 works fine.

**Comment by knightsamar** [comment]
On 783b53f with packer v 1.10 and qemu v 2.6.0, i have now managed to go beyond the error without doing any modifications!

https://pastebin.osuosl.org/107256/

Would try the workaround mentioned in https://bugzilla.redhat.com/show_bug.cgi?id=1440619 next.

**Comment by ramereth** [comment]
@knightsamar We're running qemu-kvm-ev-2.9.0-16.el7_4.5.1.ppc64le which should have the fix mentioned in that bug included.

**Comment by knightsamar** [comment]
@ramereth I had tested by manually installing 2.6.0 on openpower8.
2.9.0 fixes the TCE Table bug yes, but also introduces a new bug which we don't have a solution for, yet.

**Comment by ramereth** [comment]
@knightsamar I did some digging and I suspect this commit qemu/qemu@720b8dc052ae2783d9b8cc2263eccc5bcec04418 might be what we're hitting. It seems they changed that from a warning to an error. Possibly related is hashicorp/packer#4774 which mentions the ability to mount the cdrom image using a different method.

**Comment by knightsamar** [comment]
So finally this:

``
/usr/libexec/qemu-kvm -m 2048M -vnc 0.0.0.0:33 -name packer-ubuntu-16.04-ppc64le-openstack -machine type=pseries,accel=kvm -netdev user,id=user.0,hostfwd=tcp::3652-:22 -boot strict=on -drive file=packer-ubuntu-16.04-ppc64le-openstack/packer-ubuntu-16.04-ppc64le-openstack,if=virtio,cache=writeback,discard=ignore,format=qcow2 -drive id=ubuntu,if=none,snapshot=on,file=/home/alfred/workspace/packer-templates/packer_cache/a2be7cab6d5c514fb2a6f432427b087a53167d37ae76c594b60b2ed550f896ef.iso,index=2,media=cdrom -device virtio-net,netdev=user.0
``

seems to start the machine without error. Now to make packer accept and give those params would be another challenge as reported in that bug ... :(

**Comment by ramereth** [comment]
Is this ready to be finally merged?

**Comment by knightsamar** [comment]
@ramereth so I started a new PR with all things from this branch except for the ubuntu templates. #39
